### DISCUSSION TITLE
Multiple Folder for Models?

### DISCUSSION THREAD

**Discussion by radry** [vladmandic/sdnext#1213] [General]
Is it possible to define multiple folders for models? I'm on windows, so no easy symlinks...
Background is I want my main model on an SSD but less used ones on HDD.

**Comment by SoraEnkai** [comment]
if you're on windows 11, you can do symlinks in `cmd`
`mklink /J "target folder" "original folder"`

Should also work on win 10, but you'll need to run cmd as admin.

More info https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/mklink

**Reply by Razunter** [reply]
Also there is this tool: https://schinagl.priv.at/nt/hardlinkshellext/linkshellextension.html

**Reply by radry** [reply]
Thanks, but the downside of using symlink in windows is, the link is not shown in the explorer. It's an invisible link that you have to remember exists.

**Reply by Razunter** [reply]
Symlinks have an icon, hardlinks do not.
![symlink](https://github.com/vladmandic/automatic/assets/362900/42988349-b7a1-40f7-9cca-babe81f3e994)

**Reply by radry** [reply]
I don't see an icon. Do I have to create the folder it links to first or not? This is very confusing.  
I opened cmd in the Stable Diffusion/models folder and run:
`mklink /J \SDmodels Z:\SDmodels`

Regardless if I create "SDmodels" Folder inside models Folder first or not, the link is created but there is no icon.

I'm on Windows 10 

edit: SD doesn't find the models but when I create the link cmd says link was created?

**Reply by brknsoul** [reply]
Yeah, I was gonna say this; directory symlinks and directory junctions have the shortcut arrow on their folders.

Hardlinks don't, but that's because they're hardlinks. So long as one hardlink is pointing to file data, that data won't get marked for overwriting.

![image](https://github.com/vladmandic/automatic/assets/1969381/559d930e-3a73-40b0-a06f-8bc49268a827)

**Reply by brknsoul** [reply]
> edit: SD doesn't find the models but when I create the link cmd says link was created?

You created the junction in the root folder of that drive. Remove the `\` if you want to create it in the current folder.
`mklink /J SDmodels Z:\SDmodels`

**Comment by brknsoul** [comment]
~You could also straight up symlink the files themselves;~
~`mklink "C:\automatic\models\Stable diffusion\modelname.safetensors" "D:\LesserUsedModels\modelname.safetensors"`~
~C: being your install of SD.Next on an SSD and D: being the HDD where your lesser used models are stored.~


Linking folders inside `Stable-diffusion\` is a better method.

**Reply by vladmandic** [reply]
linking individual models can get messy. if you want to have multiple models folders, cleanest is to leave main models folder as-is and inside it create symlinks that point to other locations. so each location becomes a subfolder inside models folder.

**Reply by brknsoul** [reply]
If I had a folder, eg: `automatic\models\Stable diffusion\SomeRandomFolder\` with model files in there, is that read?

**Reply by SoraEnkai** [reply]
yes if you use directory junction `/J` 
for example: `mklink /J "H:/targetfolder/" "C:/automatic/models/Stable diffusion/SomeRandomFolder/"` etc would make a symlink of SomeRandomFolder inside H;/targetfolder/

**Reply by brknsoul** [reply]
> yes if you use directory junction `/J` for example: `mklink /J "H:/targetfolder/" "C:/automatic/models/Stable diffusion/SomeRandomFolder/"` etc would make a symlink of SomeRandomFolder inside H;/targetfolder/

Yes, I understand that, I'm just wondering if SD.Next reads files within folders inside `Stable diffusion` or just files in `Stable diffusion`?

I know that SD.Next detects
`automatic\models\Stable diffusion\SomeModel.safetensors`
but does it detect
`automatic\models\Stable diffusion\aRandomFolder\AnotherModel.safetensors`?

**Reply by SoraEnkai** [reply]
It does read the folders, they are also shown in checkpoint selection with folder names, same for LORA/LyCORIS etc.

**Reply by brknsoul** [reply]
Ahh, cool!

![image](https://github.com/vladmandic/automatic/assets/1969381/e36fe3cc-0d8c-4b1f-8407-2cf58e6a1671)

So, yeah, symlinking folders is a better idea.

**Reply by radry** [reply]
> yes if you use directory junction `/J` for example: `mklink /J "H:/targetfolder/" "C:/automatic/models/Stable diffusion/SomeRandomFolder/"` etc would make a symlink of SomeRandomFolder inside H;/targetfolder/

Shouldn't the paths be swaped here? The first should be the target, the second the "source".
`mklink [[/d] | [/h] | [/j]] <link> <target>`
`<link>  Specifies the name of the symbolic link being created. 
<target> Specifies the path (relative or absolute) that the new symbolic link refers to.
`

**Reply by SoraEnkai** [reply]
Yeah, sorry might have worded it wrongly, but targetfolder would be the place where the symlink will be added to and second path is the source folder added to that link.
